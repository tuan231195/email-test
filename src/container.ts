import 'reflect-metadata';
import path from 'path';
import Container, { ContainerInstance } from 'typedi';
import { resolveAsyncConfigs } from 'config/async';
import { configToken } from 'src/tokens';

process.env.NODE_CONFIG_DIR = path.resolve(__dirname, 'config');

const config = require('config');

let containerPromise: Promise<ContainerInstance>;

export async function initContainer() {
	if (containerPromise) {
		return containerPromise;
	}
	containerPromise = new Promise(async (resolve, reject) => {
		try {
			const container = Container.of();
			await resolveAsyncConfigs(config);
			container.set(configToken, config);
			resolve(container);
		} catch (e) {
			reject(e);
		}
	});
	return containerPromise;
}
