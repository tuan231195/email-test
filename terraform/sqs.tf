locals {
	message_retention_period = 1209600
	max_received_count = 3
	visibility_timeout_seconds = 360
}
resource "aws_sqs_queue" "email_queue" {
	name = "email-queue"
	message_retention_seconds = local.message_retention_period
	visibility_timeout_seconds = local.visibility_timeout_seconds
	redrive_policy = jsonencode({
		deadLetterTargetArn = aws_sqs_queue.email_queue_dlq.arn,
		maxReceiveCount = local.max_received_count
	})
}

resource "aws_sqs_queue" "email_queue_dlq" {
	name = "email-queue-dlq"
	message_retention_seconds = local.message_retention_period
}
