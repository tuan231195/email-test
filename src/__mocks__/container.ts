import 'reflect-metadata';
import { mockConfig } from 'src/test/config';
import { LoggingService } from 'src/core/logging/services/logging.service';
import { SqsService } from 'src/aws/sqs';
import { createSpyObj } from 'jest-createspyobj';
import { configToken } from 'src/tokens';
import { EmailService } from 'src/email/services/email.service';

const instances = new Map();
const mockSqs = createSpyObj(SqsService);
const mockEmailService = createSpyObj(EmailService);

instances.set(configToken, mockConfig({}));
instances.set(EmailService, mockEmailService);
instances.set(LoggingService, console);
instances.set(SqsService, mockSqs);

export function initContainer() {
	return Promise.resolve({
		get(clazz) {
			return instances.get(clazz);
		},
	});
}
