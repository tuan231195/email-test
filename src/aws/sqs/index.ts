import { SQSRecord } from 'aws-lambda';
import { Inject, Service } from 'typedi';
import AWS, { SQS } from 'aws-sdk';
import { chunk } from 'lodash';
import { LoggingService } from 'src/core/logging/services/logging.service';
import { configToken } from 'src/container';
import { IConfig } from 'config';

@Service()
export class SqsService {
	private sqs: SQS;
	constructor(
		@Inject() private readonly loggingService: LoggingService,
		@Inject(configToken) private readonly config: IConfig
	) {
		this.sqs = new AWS.SQS({
			region: config.get('aws.region'),
		});
	}

	async sendMessage({
		queueUrl,
		message,
		messageGroupId = undefined,
	}: {
		queueUrl: string;
		message: any;
		messageGroupId?: string;
	}) {
		await this.sqs
			.sendMessage({
				QueueUrl: queueUrl,
				MessageBody: JSON.stringify(message),
				MessageGroupId: messageGroupId,
			})
			.promise();
	}

	async processMessages({
		queueUrl,
		messages,
		handler,
	}: {
		queueUrl: string;
		messages: SQSRecord[];
		handler: (message: SQSRecord) => any;
	}) {
		// we are fine with processing multiple message groups in parallel. However, messages in the same group must be processed sequentially
		const processedResults = await Promise.all(
			messages.map(
				async message =>
					await this.processSingleMessage({ message, handler })
			)
		);

		const hasPartialFailures = processedResults.some(
			result => !result.success
		);
		if (!hasPartialFailures) {
			// no partial failures, let AWS handles message deletion
			return;
		}
		const successMessages = processedResults
			.filter(result => result.success)
			.map(result => result.message);
		const failedMessageIds = processedResults
			.filter(result => !result.success)
			.map(result => result.message.messageId);
		await this.deleteMessages({ queueUrl, messages: successMessages });
		this.loggingService.error(
			'The following message failed in this batch ' + failedMessageIds
		);
		throw new Error('Message Batch not processed successfully');
	}

	async processSingleMessage({
		message,
		handler,
	}: {
		message: SQSRecord;
		handler: (message: SQSRecord) => any;
	}) {
		try {
			const result = await handler(message);
			return { success: true, result, message };
		} catch (error) {
			return { success: false, error, message };
		}
	}

	async deleteMessages({
		queueUrl,
		messages,
	}: {
		queueUrl: string;
		messages: SQSRecord[];
	}) {
		if (!messages.length) {
			return;
		}
		// sqs can only delete up to 10 messages at a time
		const messageChunks = chunk(messages, 10);
		for (const messageChunk of messageChunks) {
			await this.sqs
				.deleteMessageBatch({
					Entries: messageChunk.map(message => ({
						Id: message.messageId,
						ReceiptHandle: message.receiptHandle,
					})),
					QueueUrl: queueUrl,
				})
				.promise();
		}
	}
}
