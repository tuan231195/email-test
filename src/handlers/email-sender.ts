import { initContainer } from 'src/container';
import { error, ok } from 'src/core/http/utils/response';
import { Email } from 'src/email/models/email';
import { validate } from 'yup-decorator';
import { sanitizeHtml } from 'src/utils/html/sanitize';
import { CustomValidationError } from 'src/core/errors/custom-validation-error';
import { SqsService } from 'src/aws/sqs';
import { configToken } from 'src/tokens';

export async function handler(event) {
	try {
		const container = await initContainer();
		const email = await verifySendInformation(JSON.parse(event.body));
		const sqsService = container.get(SqsService);
		const config = container.get(configToken);

		await sqsService.sendMessage({
			queueUrl: config.get('email.queueUrl'),
			message: email,
		});

		return ok({ message: 'Email sent successfully' });
	} catch (e) {
		return error(e);
	}
}

async function verifySendInformation(emailContent): Promise<Email> {
	try {
		await validate({
			schemaName: Email,
			object: emailContent,
			options: {
				abortEarly: false,
			},
		});
		return {
			...emailContent,
			body: sanitizeHtml(emailContent.body),
		};
	} catch (e) {
		throw new CustomValidationError(e);
	}
}
