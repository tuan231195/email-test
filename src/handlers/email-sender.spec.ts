import { handler } from './email-sender';
import testCase from '../email/adapters/test/test.case';
import { initContainer } from '../container';
import { SqsService } from '../aws/sqs';

jest.mock('src/container');

describe('Testing email.handler', () => {
	describe('Test input validation', () => {
		test('should return an error when the there is no sender details', async () => {
			const result = await handler({
				body: JSON.stringify({
					...testCase,
					from: {},
				}),
			});
			expect(result.statusCode).toEqual(400);
			expect(result).toMatchSnapshot();
		});

		test('should return an error when the there is no recipients', async () => {
			const result = await handler({
				body: JSON.stringify({
					...testCase,
					to: [],
				}),
			});
			expect(result.statusCode).toEqual(400);
			expect(result).toMatchSnapshot();
		});

		test('should return an error when sender email is incorrect', async () => {
			const result = await handler({
				body: JSON.stringify({
					...testCase,
					from: {
						name: 'a',
						email: 'vdtn359',
					},
				}),
			});
			expect(result.statusCode).toEqual(400);
			expect(result).toMatchSnapshot();
		});

		test('should return an error when recipient name is missing', async () => {
			const result = await handler({
				body: JSON.stringify({
					...testCase,
					from: {
						name: 'a',
						email: 'vdtn359@gmail.com',
					},
					to: [
						{
							email: 'vdtn359@gmail.com',
						},
					],
				}),
			});
			expect(result.statusCode).toEqual(400);
			expect(result).toMatchSnapshot();
		});

		test('should return an error when recipient email is incorrect', async () => {
			const result = await handler({
				body: JSON.stringify({
					...testCase,
					from: {
						name: 'a',
						email: 'vdtn359@gmail.com',
					},
					to: [
						{
							name: 'a',
							email: 'vdtn359',
						},
					],
				}),
			});
			expect(result.statusCode).toEqual(400);
			expect(result).toMatchSnapshot();
		});

		test('should sanitize body', async () => {
			const result = await handler({
				body: JSON.stringify({
					...testCase,
					body: '<script>bad</script>Test',
				}),
			});
			expect(result.statusCode).toEqual(200);
			expect(result).toMatchSnapshot();
			const container = await initContainer();
			const sqsService = container.get(SqsService);
			expect(sqsService.sendMessage).toHaveBeenCalled();
			expect(
				(sqsService.sendMessage as any).mock.calls
			).toMatchSnapshot();
		});
	});
});
