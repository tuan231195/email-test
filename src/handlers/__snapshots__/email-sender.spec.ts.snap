// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing email.handler Test input validation should return an error when recipient email is incorrect 1`] = `
Object {
  "body": "{\\"errors\\":[{\\"field\\":\\"to[0].email\\",\\"message\\":\\"Not a valid email\\"}]}",
  "headers": Object {
    "Access-Control-Allow-Origin": "*",
    "Content-Type": "application/json",
  },
  "statusCode": 400,
}
`;

exports[`Testing email.handler Test input validation should return an error when recipient name is missing 1`] = `
Object {
  "body": "{\\"errors\\":[{\\"field\\":\\"to[0].name\\",\\"message\\":\\"Recipient name is required\\"}]}",
  "headers": Object {
    "Access-Control-Allow-Origin": "*",
    "Content-Type": "application/json",
  },
  "statusCode": 400,
}
`;

exports[`Testing email.handler Test input validation should return an error when sender email is incorrect 1`] = `
Object {
  "body": "{\\"errors\\":[{\\"field\\":\\"from.email\\",\\"message\\":\\"Not a valid email\\"}]}",
  "headers": Object {
    "Access-Control-Allow-Origin": "*",
    "Content-Type": "application/json",
  },
  "statusCode": 400,
}
`;

exports[`Testing email.handler Test input validation should return an error when the there is no recipients 1`] = `
Object {
  "body": "{\\"errors\\":[{\\"field\\":\\"to\\",\\"message\\":\\"Email recipient is required\\"}]}",
  "headers": Object {
    "Access-Control-Allow-Origin": "*",
    "Content-Type": "application/json",
  },
  "statusCode": 400,
}
`;

exports[`Testing email.handler Test input validation should return an error when the there is no sender details 1`] = `
Object {
  "body": "{\\"errors\\":[{\\"field\\":\\"from.name\\",\\"message\\":\\"Recipient name is required\\"},{\\"field\\":\\"from.email\\",\\"message\\":\\"Recipient email is required\\"}]}",
  "headers": Object {
    "Access-Control-Allow-Origin": "*",
    "Content-Type": "application/json",
  },
  "statusCode": 400,
}
`;

exports[`Testing email.handler Test input validation should sanitize body 1`] = `
Object {
  "body": "{\\"message\\":\\"Email sent successfully\\"}",
  "headers": Object {
    "Access-Control-Allow-Origin": "*",
    "Content-Type": "application/json",
  },
  "statusCode": 200,
}
`;

exports[`Testing email.handler Test input validation should sanitize body 2`] = `
Array [
  Array [
    Object {
      "message": Object {
        "bcc": Array [
          Object {
            "email": "vdtn359@gmail.com",
            "name": "Nguyen",
          },
        ],
        "body": "Test",
        "cc": Array [
          Object {
            "email": "vdtn359@uowmail.edu.au",
            "name": "TN",
          },
        ],
        "from": Object {
          "email": "vdtn359@gmail.com",
          "name": "Tuan Nguyen",
        },
        "subject": "Test subject",
        "to": Array [
          Object {
            "email": "vdtn359@uowmail.edu.au",
            "name": "Tuan",
          },
        ],
      },
      "queueUrl": undefined,
    },
  ],
]
`;
